### ISSUE TITLE
Potential issues in Metric for empty tensor

### ISSUE THREAD

**Issue by pkdogcom** [pytorch/ignite#357]
I encountered an issue when my output_transform given to Metric perform certain filtering (e.g. when dealing with multi-dataset training and want to have separate metrics for different datasets) and the transformed (filtered) output happen to be empty. In this case metric like Accuracy will raise error because of unsqueezing empty tensor. 

I think the base Metric class may need to perform a quick filtering in iteration_completed method to avoid passing empty tensor into actual update implementation of subclass.

**Comment by vfdev-5** [comment]
@pkdogcom thanks for your feedback! I have doubt on what you call multi-dataset trainin, could you please detail this. 
I see your point on filtering data during metrics computing but I'm not sure that this would be a correct lib usage.

**Comment by pkdogcom** [comment]
For example, I may want to train a single multi-task model (with shared conv feature extractor and multiple output heads) which perform two different types of classification, say classifying a person's gender as well as types of clothes they are wearing, while each of the task can be trained by a dedicated dataset. 
Since each dataset only contains label for one task, during back-propagation as well as evaluation I'll need to use label mask and only calculate loss/metrics on available labels for each sample, which is implemented as output_transform in Metric. The issue I mentioned happens when the batch of data happen to only contain samples from one dataset and thus the masked/filtered output of the batch may be empty tensor for the other task/loss/metric. 
The current way to resolve it is shuffle the data during evaluation and use a larger batch to make sure each batch contain at least one sample for every dataset or task. I was wondering if a better way would be to let each metric handle this corner case and just perform no-op instead (may be printing a warning message for those who accidentally pass empty tensor to metric)

**Comment by vfdev-5** [comment]
@pkdogcom I see better your problem. There are probably at least two solutions for this.
The first one, as your problem is rather specific, you can use a sort of wrapper over `Metric` class as `RunningAverage` to filter out empty output : 
```python
import torch 

y_true_batches = [
    [0, 1, 2, 3, 0],
    [1, 2, 3, 0, 1],
    [2, 3, 0, 0, 1]
]
y_pred_batches = [
    torch.rand(5, 4),
    torch.rand(5, 4),
    torch.rand(5, 4)
]
mask_gender_batches = [
    [0, 1, 1, 0, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 1, 1, 1]
]

y_true_batch_iter = iter(y_true_batches)
y_pred_batch_iter = iter(y_pred_batches)
mask_gender_batch_iter = iter(mask_gender_batches)

def update_fn(engine, _):
    y_true_batch = next(y_true_batch_iter)
    y_pred_batch = next(y_pred_batch_iter)
    mask_gender_batch = next(mask_gender_batch_iter)
    
    return y_pred_batch, \
            torch.Tensor(y_true_batch).type(torch.LongTensor), \
            torch.Tensor(mask_gender_batch).type(torch.ByteTensor)


from ignite.engine import Engine


evaluator = Engine(update_fn)

def gender_output_transform(output):
    y_pred, y, mask_gender = output
    y_gender = y[mask_gender]
    y_pred_gender = y_pred[mask_gender]    
    return y_pred_gender, y_gender

from ignite.metrics.metric import Metric


class EmptyTargetFiltering(Metric):
    
    def __init__(self, metric):
        self.metric = metric
        super(EmptyTargetFiltering, self).__init__(output_transform=self.metric._output_transform)
            
    def reset(self):
        self.metric.reset()
        
    def update(self, output):
        # check if output is empty
        y_pred, y_true = output
        if len(y_pred) > 0 and len(y_true) > 0:
            self.metric.update(output)

    def compute(self):
        return self.metric.compute()

acc = Accuracy(output_transform=gender_output_transform)
gender_acc = EmptyTargetFiltering(acc)
gender_acc.attach(evaluator, "gender_accuracy")

evaluator.run([0, 1, 2], max_epochs=1)

evaluator.state.metrics["gender_accuracy"]
```

The second way is to split model into two model with shared features, setup the multi-task dataset as `(x_gender, x_clothes), (y_gender, y_clothes) = batch` and thus compute metrics as usual. The drawback is that you have perform two forward passes instead of one.
